### PR TITLE
feat(firestore): add the `Filter` class and support for the OR query

### PR DESCRIPTION
**Pull request by Lyokone**
## Description

This PR adds support for Filter as a new argument of `where()`.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database](https://github.com/flutter/flutter/issues). Indicate, which of these issues are resolved or fixed by this PR. Note that you'll have to prefix the issue numbers with flutter/flutter#.*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Review by russellwheatley — APPROVED** [review]

File: packages/cloud_firestore/cloud_firestore/lib/src/query.dart [hunk 1]
```diff
@@ -603,7 +607,13 @@ class _JsonQuery implements Query<Map<String, dynamic>> {
     Iterable<Object?>? whereNotIn,
     bool? isNull,
   }) {
-    _assertValidFieldType(field);
+    _assertValidFieldType(fieldOrFilter);
+
+    if (fieldOrFilter is Filter) {
```
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore/lib/src/query.dart:612]: Shouldn't there be an assert which prevents the usage of both Filter + isEqualTo/co? It sounds redundant

**Review by rrousselGit — APPROVED** [review]

File: packages/cloud_firestore/cloud_firestore/lib/src/query.dart (hunk 1 above)
Comment (Lyokone) [reply packages/cloud_firestore/cloud_firestore/lib/src/query.dart:612]: Nice catch, I've added it :)

**Comment by Lyokone** [comment]
@kevinthecheung We added support for Filter and the OR query with this PR. We cannot update the documentation for Firestore, the API is pretty similar to the Java one, here are the 2 examples from the documentation translated in Dart:
```dart
    final query = collection.where(
      Filter.or(
        Filter('capital', isEqualTo: true),
        Filter('population', isGreaterThanOrEqualTo: 1000000),
      ),
    );
```

```dart
    final query = collection.where(
      Filter.and(
        Filter('state', isEqualTo: 'CA'),
        Filter.or(
          Filter('capital', isEqualTo: true),
          Filter('population', isGreaterThanOrEqualTo: 1000000),
        ),
      ),
    );

```

**Comment by Nico04** [comment]
That's awesome ! Do you now in which release it will be available ? When ? Thanks :)

**Comment by Lyokone** [comment]
It's already available with cloud_firestore 4.5.0 :)

**Comment by bahadurh** [comment]
Is there any documentation added yet? im not able to use the filter classes for some reasons, does anybody faces this issue? ps: i have the updated 4.5.0 but smh cant use the code snippet :
    final query = collection.where(
      Filter.and(
        Filter('state', isEqualTo: 'CA'),
        Filter.or(
          Filter('capital', isEqualTo: true),
          Filter('population', isGreaterThanOrEqualTo: 1000000),
        ),
      ),
    );

as the Filter class cant be imported.

**Comment by Nico04** [comment]
> Is there any documentation added yet? im not able to use the filter classes for some reasons, does anybody faces this issue? ps: i have the updated 4.5.0 but smh cant use the code snippet :
> 
> ```
>     final query = collection.where(
>       Filter.and(
>         Filter('state', isEqualTo: 'CA'),
>         Filter.or(
>           Filter('capital', isEqualTo: true),
>           Filter('population', isGreaterThanOrEqualTo: 1000000),
>         ),
>       ),
>     );
> ```
> 
> as the Filter class cant be imported.

Weird, cause for me it's working great. Maybe a pub cache issue ?

**Comment by Lyokone** [comment]
@bahadurh We are waiting on the update to be pushed to the official website, but you can check the [Kotlin code](https://firebase.google.com/docs/firestore/query-data/queries#kotlin+ktx_14) while waiting, it's pretty similar.

You can probably restart your IDE or manually run `flutter clean` and `flutter pub get` to be sure to have everything ready.

**Pull request merged**